engine/consensus/compliance/engine.go

```diff
@@ -19,98 +17,126 @@ import (
 	"github.com/onflow/flow-go/module/metrics"
 	"github.com/onflow/flow-go/module/trace"
 	"github.com/onflow/flow-go/network"
-	"github.com/onflow/flow-go/state"
 	"github.com/onflow/flow-go/state/protocol"
 	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/fifoqueue"
 	"github.com/onflow/flow-go/utils/logging"
 )
 
-// Engine is the consensus engine, responsible for handling communication for
-// the embedded consensus algorithm.
+type Event struct {
+	OriginID flow.Identifier
+	Msg      interface{}
+}
+
+type (
+	EventSink chan *Event // Channel to push pending events
+)
+
+// defaultBlockQueueCapacity maximum capacity of block proposals queue
+const defaultBlockQueueCapacity = 10000
+
+// defaultVoteQueueCapacity maximum capacity of block votes queue
+const defaultVoteQueueCapacity = 10000
+
+// Engine is a wrapper struct for `Core` which implements consensus algorithm.
+// Engine is responsible for handling incoming messages, queueing for processing, broadcasting proposals.
 type Engine struct {
-	unit     *engine.Unit   // used to control startup/shutdown
-	log      zerolog.Logger // used to log relevant actions with context
-	metrics  module.EngineMetrics
-	tracer   module.Tracer
-	mempool  module.MempoolMetrics
-	spans    module.ConsensusMetrics
-	me       module.Local
-	cleaner  storage.Cleaner
-	headers  storage.Headers
-	payloads storage.Payloads
-	state    protocol.MutableState
-	con      network.Conduit
-	prov     network.Engine
-	pending  module.PendingBlockBuffer // pending block cache
-	sync     module.BlockRequester
-	hotstuff module.HotStuff
+	unit             *engine.Unit
+	log              zerolog.Logger
+	mempool          module.MempoolMetrics
+	metrics          module.EngineMetrics
+	me               module.Local
+	headers          storage.Headers
+	payloads         storage.Payloads
+	tracer           module.Tracer
+	state            protocol.State
+	prov             network.Engine
+	core             *Core
```

Comment:
Another benefit of taking `Core` as input is that it could make the code testable.

In the test case of the Engine, we could make `Core` as a mock, and verify the concurrency behavior by sending blocks and votes to the Engine, and check in what order did the mocked `Core` receive those messages/events.

So we don't need to duplicate the behavior tests to the Engine, but only tests the Engine's own responsibility which is to ensure the `Core` receives one event at a time.

Reply:
Those are just for easier access, some of those are only for engine and aren't part of core. And `Core` is taken as input as well(are you asking to make core the only input?)

Reply:
>are you asking to make core the only input

yes

Reply:
Unfortunately, I don't understand the reasoning in your [comment](https://github.com/onflow/flow-go/pull/406#discussion_r580566218) @zhangchiqing. I think the Engine uses each of these imports in its own methods. 
* We _could_ move them all into `compliance.Core` and then have `compliance.Engine` access internal fields of `Core`. But what is the benefit?
   * In my mind, a clear separation of concern is very desirable. Hence, I would personally prefer to only use exported methods from `Core`, but no internal fields. Then, we can easily abstract the inner working of `Core` through an interface (like what we have done with  `EventLoop` and `EventHandler`), apply wrappers, further modularize the code, etc 
   * I see little benefits in saving a few imports for the `Engine` the cost of blurring the separation of concern between `matching.Engine`  and `matching.Core`

Reply:
I was suggesting to make the `Core` in cmd/consensus/main with all the dependencies, and passing it to the Engine, so that the `Core` is the only input of the `Engine`

Reply:
But core doesn't use all of those, I think accessing every member using `e.core.smth` is not that good, I am doing it only with hotstuff(there is a good reason for that) and I would better hold references to fields that are used in both places then accessing them from core.